### PR TITLE
Hogan-style template inheritance

### PR DESCRIPTION
**Pull request by gasche**
This PR implements Hogan-style "template inheritance".

A simple way to think of this feature is that partials behave like functions (expanded at a "callsite"), and template inheritance adds functions that take arguments: at the callsite of a partial, we can pass template fragments as named arguments. For example, running `mustache empty.json caller.mustache` with

function.mustache:
```
This template is parametrized over {{$arg}}{{/arg}}.
```

caller.mustache:
```
{{<function}}{{$arg}}an argument{{/arg}}{{/function}}
```

renders as `This template is parametrized over an argument.`

Simple partials `{{>foo}}` allow to place the content a template defined elsewhere at a leaf of a template file. Partial-with-arguments `{{<foo}}{{$arg}}..{{/arg}}{{/foo}}` allow to place the content of a template *around* parts of the current template.

This feature is specified in a pull-request to the "Mustache specs" repository:

  https://github.com/mustache/spec/pull/75

and supported by many Mustache implementations:

- [hogan.js](https://github.com/twitter/hogan.js)
- [mustache.php](https://github.com/bobthecow/mustache.php)
- [mustache.java](https://github.com/spullara/mustache.java)
- [GRMustache](https://github.com/groue/GRMustache) (Obj-C) and [GRMustache.swift](https://github.com/groue/GRMustache.swift) (Swift)
- [Text::Caml](https://github.com/vti/text-caml) (Perl)
- [hxmustache](https://github.com/nadako/hxmustache) (Haxe)
- [MuttonChop](https://github.com/Danappelxx/MuttonChop) (Swift)

I have a use for the feature myself. I don't really like the syntax that was chosen, but I think there is real value in using the same syntax and semantics as other implementations. The proposed implementation was tested against the tests from the proposed specification.

#### Compatibility

The ASTs of the Mustache implementation have changed: the `Partial` constructor takes an extra parameter `params` (`None` for usual partials, `Some [...]` for partials with arguments), and there is a new `Partial_param` constructor. This breaks compatibility.

(Having to change two ASTs at once, for the with-location and the no-location version, proved a bit painful when implementing this feature. I might try to factor the two ASTs using a parametrized type.)

#### Standalone fun

The tests in the specification revealed a shortcoming of the current "standalone handling" implementation: it does not support cases where several standalone tags are on the same line, for example `{{#foo}}{{#bar}}`. This was not tested in existing specifications, but there is a test in the inheritance specification that checks that `{{<foo}}{{$param}}` is handled as a standalone line. To pass this test, I had to change the standalone-handling implementation (I tried several things). The new implementation passes all the previous tests, and the template-inheritance tests.

Partial-with-arguments introduce standalone blocks with an open and a close tag (`{{<foo}}...{{/foo}}` instead of just `{{>foo}}`). This is difficult for a lexer-postprocessing approach to handle, and several cases are in fact not considered as "standalone lines" when they arguably should:

```
{{<foo}}{{$arg}}Some parameter {{value}}.{{/arg}}{{/foo}}
{{<foo}}{{$arg}}Some parameter {{value}}.{{/arg}}
{{<foo}}{{$arg1}}Some parameter {{value}}.{{/arg1}}{{$arg2}}
```

I have an idea of how those cases could be fixed, but this is more subtle code to handle corner cases, and conceptually ugly code. Instead I propose to just forget about these corner cases. There is an easy workaround (to get the expected whitespace behavior in the output), which is to always have `{{<foo}}` and `{{/foo}}` on their own lines.

### CODE REVIEW

File: lib_test/dune
Comment (rgrinberg) [inline lib_test/dune:10]: I'd turn this into a glob to ease maintenance `(glob_files ../specs/*.json)`.

File: specs/inheritance.json
Comment (rgrinberg) [inline specs/inheritance.json:1]: Can you pretty-print this file perhaps? Or is the minification required for the test?

File: specs/inheritance.yml
Comment (rgrinberg) [inline specs/inheritance.yml:1]: Is this yaml file added for the purpose of documentation?

File: lib_test/test_mustache.ml
Comment (rgrinberg) [inline lib_test/test_mustache.ml:112]: A cram test would make it easier to demonstrate the fix.

File: bin/test/errors/parsing-errors.t
Comment (rgrinberg) [inline bin/test/errors/parsing-errors.t:115]: Does this style not apply to variables? E.g.  `{{$ foo }}..{{/ foo}}`

**Comment by gasche** [comment]
Thanks @rgrinberg for the review! (I hope that I am not disturbing you in some sort of well-deserved vacation with my PRs. Please feel free to take a few days to handle them.)

A point worth noting is that this PR strengthens the OCaml version requirement from 4.06 to 4.08. There is no very strong reason for this, I just happened to have a use for `List.filter_map` which was added in 4.08 (and this allowed me to get rid of Mustache's reimplementation of `Option.map`); if this stronger requirement was an issue I wouldn't mind rewriting the code to keep working on 4.06. 

(Debian stable is currently at 4.05, if I read correctly.)

File: specs/inheritance.yml
Comment (gasche) [reply specs/inheritance.yml:1]: (This is also related to your other question on `.json` minification). Specs in the [official repository](https://github.com/mustache/spec) are written in YAML format as the reference, and then [programmatically](https://github.com/mustache/spec/blob/master/Rakefile#L12-L27) rendered into JSON to be usable from people outside the YAML cult.

The existing spec files in the `specs/` directory of ocaml-mustache exactly follow this format: both versions wered copied over, the JSON version is used by the tool and the YAML version is read by humans when a test fails. The import I did here follows the same approach.

I think that it is okay to keep the current (non-ideal) workflow. In any case, we should keep the YAML files around, because they are the reference format, so we want people to be able to "diff" them against the current version of the spec to see if we missed some changes. (The spec repository currently appears to be unmaintained, but oh well.)

We could probably add some tooling of our side to re-render the JSON files from the YAML file with human-friendly pretty-printing. (One fun aspect of it is that the JSON specs from the upstream repo are [currently out of date](https://github.com/mustache/spec/pull/113), so doing this refresh would in fact change slightly the tests we are running. Who knows, they may even catch a mistake and start failing.) I would propose to wait until there is some `yaml` library used here (possibly from #45)  to do this.

File: lib_test/test_mustache.ml
Comment (gasche) [reply lib_test/test_mustache.ml:119]: There is a difficulty in terms of test organization, though: if we want to split the binary as a separate package, I probably shouldn't write a cram test using the binary in the library tests, as this would introduce a suspicious circular-like dependency (lib{test} -> bin -> lib). (For now the cram tests in `bin/test` are more like integration tests for the binary behavior, rather than arbitrary spec tests.)

I could write a cram test using the library only. For now I went the easy way of just using 0-indented string literals with real newlines instead of `\n` escapes, to make the current test easier to read and understand.

File: lib_test/dune
Comment (gasche) [reply lib_test/dune]: Thanks! Done.

File: bin/test/errors/parsing-errors.t
Comment (gasche) [reply bin/test/errors/parsing-errors.t:115]: It does, I just added test cases for `{{$foo}}` and `{{<foo}}`. Feel free to recommend improvements to the wording.

File: specs/inheritance.yml
Comment (rgrinberg) [reply specs/inheritance.yml:1]: Alright. We should note this in a ticket then. Using a yaml library for the tests sounds fine as well.

File: bin/test/errors/parsing-errors.t
Comment (rgrinberg) [reply bin/test/errors/parsing-errors.t:115]: Sorry, what I meant was that `pp_partial_param` isn't consistent with this printing style. It's still printing `{{$foo}}` instead of `{{$ foo }}`.

**Review by rgrinberg — APPROVED** [review]
I don't use 4.06 or 4.07 so I have no problem with dropping support.

**Comment by gasche** [comment]
This needs a Changes entry, and I am planning to wait until #59 and #60 are merged to complete the tool manpage with a section on inheritance.

**Comment by Drup** [comment]
This is interesting, I think this was one of the missing feature that made mustache not-so-suitable for https://discuss.ocaml.org/t/choice-of-template-language-for-tyxml/1863 . Maybe I could try again....

**Comment by gasche** [comment]
### Meta

I spent a few days stuck in a hole, trying to get a version of this patch that would:

1. indent template parameters in "the right way"
2. respect the specification and be compatible with other implementations
3. have readable code

I have a version that does (3) but not (2): I think it handles whitespace better than other mustache implementations, but obviously this results in different (nicer) outputs on some templates. I think that consistency is very important, so I gave up on this approach for now. (In case people would be curious, which I don't recommend, I saved it [here](https://github.com/gasche/ocaml-mustache/commits/standalone)).

I pushed a commit that does (1) and (2), but not (3): the code is not much worse than the current implementation of whitespace handling, but it still adds a bit of complexity to that obscure part of the codebase and I'm not proud of it. But I really wanted my output to be indented nicely, so I went with it and included it in the PR.

### Indentation of template parameters

There are two components that are important for nice indentation of template parameters:
1. The indentation of a parameter when rendered should be the indentation of its use-site *minus* the indentation of its definition-site.
2. The indentation of a parameter should be measured as the indentation of its *content*, when the content is itself indented.

#### Substracting definition-site indentation

Consider:

template.mustache:
```
{{#should-use-layout1}}
  {{<layout1}}
    {{$param}}
    foo
    bar
    {{/param}}
  {{/layout1}}
{{/should-use-layout1}}
```

layout1.mustache:
```
The multiline parameter is:
  {{$param}}{{/param}}
```

In this example, what we expect is that `{{$param}}{{/param}}` will get substituted by `foo\nbar\n`. Because at the use-site (in layout1) the parameter occurrence is 2-indented, the actual output should be 2-indented: `  foo\n  bar\n` (this is how partial work today). But if you naively include just the parameter-definition lines, which are itself 4-indented, you get a 6-indented output, which looks ugly!

When printing a line coming from a parameter, the indentation should be the use-site indentation *minus* the definition-site indentation of the parameter.

(This means that some line sometimes get printed with *negative* indentation, when the definition-site indentation is higher as in this example. This means that the corresponding number of spaces is skipped at the beginning of the line.)

This adds a bit of code to the implementation, but this code is not particular difficult/complex so I guess it is fine.

#### Measuring a parameter indentation

In languages that encourage indentation (typically HTML), a typical definition of a parameter would look like

```
{{<layout}}
  {{$content}}
    <h1>...</h1>
    <p>...</p>
  {{/content}}
{{/layout}}
```

If you consider that the `content` parameter is 2-indented (because the `{{$content}}` tag is 2-indented), the actual content which is 4-indented at definition site will get 2 extra indents at each use-site.

Instead the present implementation considers that the indentation of a parameter is defined as follows:
- if the parameter tag is not followed by a newline, take the indentation of the tag itself
- if the parameter tag is followed by a newline, take the indentation of the next line after it

(Consistently with the Mustache specification, if the parameter tag is not "standalone" (there are non-whitespace characters before it, then no indentation is counted at all and the output is very ugly. Partials already work that way.)

This change is a bit more annoying because it adds a "lookahead" logic (looking into the next line) to the kludge in the lexer that deals with standalone tokens.

**Comment by gasche** [comment]
I have also added a Changes entry, and some documentation in the tool manpage.

I am satisfied with the current state of the PR and believe it is mergeable, but I'll leave it around for a few days in case people have feedback on the indentation stuff.

**Review by rgrinberg — APPROVED** [review]

**Comment by gasche** [comment]
Thanks! I rebased and plan to merge when the CI comes back.

(This feature is kind of giving me the cold feet because it is large and not-as-standard-as-the-other and it's a more invasive code change, but I think it's time to settle down and, as people say, "ship it" -- I do think it would make the library useful for more use-cases, as pointed out by @Drup.)

**Pull request merged**
